Investigate test failures caused by upgrading to C#7

Microsoft.Extensions.Primitives.Tests.DefaultNullOrEmpty_Concat tests fail now that we've upgraded to C#7. There are slight difference in IL, notably a change from [box] to [call]

```diff
--- 3278ee4_cs6.txt	2016-12-15 17:24:17.913045600 -0800
+++ ffb7c20_cs7.txt	2016-12-15 17:24:15.040486600 -0800
@@ -1,12 +1,12 @@
-    // [331 13 - 331 99]
-    IL_0036: ldloc.0      // expected
-    IL_0037: ldloc.0      // expected
+    // [354 7 - 354 119]
+    IL_0036: ldloc.0      // values
+    IL_0037: ldloc.0      // values
     IL_0038: newobj       instance void [Microsoft.Extensions.Primitives]Microsoft.Extensions.Primitives.StringValues::.ctor(string[])
     IL_003d: ldarg.1      // stringValues
     IL_003e: call         valuetype [Microsoft.Extensions.Primitives]Microsoft.Extensions.Primitives.StringValues [Microsoft.Extensions.Primitives]Microsoft.Extensions.Primitives.StringValues::Concat(valuetype [Microsoft.Extensions.Primitives]Microsoft.Extensions.Primitives.StringValues, valuetype [Microsoft.Extensions.Primitives]Microsoft.Extensions.Primitives.StringValues)
-    IL_0043: box          [Microsoft.Extensions.Primitives]Microsoft.Extensions.Primitives.StringValues
-    IL_0048: call         instance void [xunit.assert]Xunit.Assert::Equal<string>(class [mscorlib]System.Collections.Generic.IEnumerable`1<!!0/*string*/>, class [mscorlib]System.Collections.Generic.IEnumerable`1<!!0/*string*/>)
+    IL_0043: call         string[] [Microsoft.Extensions.Primitives]Microsoft.Extensions.Primitives.StringValues::op_Implicit(valuetype [Microsoft.Extensions.Primitives]Microsoft.Extensions.Primitives.StringValues)
+    IL_0048: call         instance void [xunit.assert]Xunit.Assert::Equal<string[]>(!!0/*string[]*/, !!0/*string[]*/)
 
-    // [333 13 - 333 44]
+    // [355 7 - 355 41]
     IL_004d: ldc.i4.0     
     IL_004e: newarr       [mscorlib]System.String
\ No newline at end of file

```


Full diff: https://gist.github.com/natemcmaster/3f1dee8dae8dc04def120516d70a7f36


Code: https://github.com/aspnet/Common/blob/ffb7c20fb22a31ac31d3a836a8455655867e8e16/test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs#L324-L338

Added note: looks like the new compiler fixes the overload resolution in the scenario `StringValues` does not implement `IEnumerable<T>`. Commenting out the `: List<string>` portion of the `StringValues` type declaration successfully compiles with the new compiler. The old compiler fails with an error:

`1.cs(21,19): error CS1503: Argument 2: cannot convert from 'Test.StringValues' to 'System.Collections.Generic.IEnumerable<string>'`

Either way, would be useful to file a bug in https://github.com/dotnet/roslyn/issues so they're aware of this.

--- a/test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs
+++ b/test/Microsoft.Extensions.Primitives.Tests/StringValuesTests.cs
@@ -321,7 +321,7 @@ namespace Microsoft.Extensions.Primitives
             Assert.Equal(expected, actual);
         }
 
-        [Theory(Skip = "Skipped to unblock CI. See https://github.com/aspnet/Common/issues/181")]
+        [Theory]
         [MemberData(nameof(DefaultOrNullStringValues))]
         [MemberData(nameof(EmptyStringValues))]
         public void DefaultNullOrEmpty_Concat(StringValues stringValues)

